src/memory.ts

```diff
@@ -52,27 +56,27 @@ namespace Il2Cpp {
                 return pointer.readDouble();
             case Il2Cpp.Type.enum.nativePointer:
             case Il2Cpp.Type.enum.unsignedNativePointer:
-                return pointer.readPointer();
+                return dereferenced;
             case Il2Cpp.Type.enum.pointer:
-                return new Il2Cpp.Pointer(pointer.readPointer(), type.class.baseType!);
+                return new Il2Cpp.Pointer(dereferenced, type.class.baseType!);
             case Il2Cpp.Type.enum.valueType:
+                // Never needs dereferencing
                 return new Il2Cpp.ValueType(pointer, type);
             case Il2Cpp.Type.enum.object:
             case Il2Cpp.Type.enum.class:
-                return new Il2Cpp.Object(pointer.readPointer());
+                return new Il2Cpp.Object(dereferenced);
             case Il2Cpp.Type.enum.genericInstance:
-                return type.class.isValueType ? new Il2Cpp.ValueType(pointer, type) : new Il2Cpp.Object(pointer.readPointer());
+                return type.class.isValueType ? new Il2Cpp.ValueType(pointer, type) : new Il2Cpp.Object(dereferenced);
             case Il2Cpp.Type.enum.string:
-                return new Il2Cpp.String(pointer.readPointer());
+                return new Il2Cpp.String(dereferenced);
             case Il2Cpp.Type.enum.array:
             case Il2Cpp.Type.enum.multidimensionalArray:
-                return new Il2Cpp.Array(pointer.readPointer());
+                return new Il2Cpp.Array(dereferenced);
         }
 
         raise(`couldn't read the value from ${pointer} using an unhandled or unknown type ${type.name} (${type.typeEnum}), please file an issue`);
     }
 
-    /** @internal */
```

Comment:
Err unrelated but could we have these exposed in the API? Bit more low level but super useful.

Reply:
That method is indeed here - if you invoke it, it works. It just isn't exposed (yet?)!